File: clients/client-python/gravitino/dto/requests/fileset_update_request.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.

from abc import abstractmethod
from dataclasses import dataclass, field

from dataclasses_json import config

from gravitino.api.file.fileset_change import FilesetChange
from gravitino.rest.rest_message import RESTRequest


@dataclass
class FilesetUpdateRequestBase(RESTRequest):
    _type: str = field(metadata=config(field_name="@type"))

    def __init__(self, action_type: str):
        self._type = action_type

    @abstractmethod
    def fileset_change(self):
        pass


class FilesetUpdateRequest:
    """Request to update a fileset."""

    @dataclass
    class RenameFilesetRequest(FilesetUpdateRequestBase):
        """The fileset update request for renaming a fileset."""

        _new_name: str = field(metadata=config(field_name="newName"))
        """The new name for the Fileset."""

        def __init__(self, new_name: str):
            super().__init__("rename")
            self._new_name = new_name

        def validate(self):
            """Validates the fields of the request.

            Raises:
                IllegalArgumentException if the new name is not set.
            """
            if not self._new_name:
                raise ValueError('"new_name" field is required and cannot be empty')

        def fileset_change(self):
            """Returns the fileset change.

            Returns:
                the fileset change.
            """
            return FilesetChange.rename(self._new_name)

    @dataclass
    class UpdateFilesetCommentRequest(FilesetUpdateRequestBase):
        """Represents a request to update the comment on a Fileset."""

        _new_comment: str = field(metadata=config(field_name="newComment"))
        """The new comment for the Fileset."""

        def __init__(self, new_comment: str):
            super().__init__("updateComment")
            self._new_comment = new_comment

        def validate(self):
            """Validates the fields of the request. Always pass."""
            pass

        def fileset_change(self):
            """Returns the fileset change"""
            return FilesetChange.update_comment(self._new_comment)

    @dataclass
    class SetFilesetPropertyRequest(FilesetUpdateRequestBase):
        """Represents a request to set a property on a Fileset."""

        _property: str = field(metadata=config(field_name="property"))
        """The property to set."""

        _value: str = field(metadata=config(field_name="value"))
        """The value of the property."""

        def __init__(self, fileset_property: str, value: str):
            super().__init__("setProperty")
            self._property = fileset_property
            self._value = value

        def validate(self):
            """Validates the fields of the request.

            Raises:
                 IllegalArgumentException if fileset_property or value are not set.
            """
            if not self._property:
                raise ValueError(
                    '"fileset_property" field is required and cannot be empty'
                )
            if not self._value:
                raise ValueError('"value" field is required and cannot be empty')

        def fileset_change(self):
            return FilesetChange.set_property(self._property, self._value)

    @dataclass
    class RemoveFilesetPropertyRequest(FilesetUpdateRequestBase):
        """Represents a request to remove a property from a Fileset."""

        _property: str = field(metadata=config(field_name="property"))
        """The property to remove."""

        def __init__(self, fileset_property: str):
            super().__init__("removeProperty")
            self._property = fileset_property

        def validate(self):
            """Validates the fields of the request.

            Raises:
                 IllegalArgumentException if fileset_property is not set.
            """
            if not self._property:
                raise ValueError(
                    '"fileset_property" field is required and cannot be empty'
                )

        def fileset_change(self):
            return FilesetChange.remove_property(self._property)

    @dataclass
    class RemoveFilesetCommentRequest(FilesetUpdateRequestBase):
        """Represents a request to remove comment from a Fileset.

        Deprecated:
            Please use `UpdateFilesetCommentRequest` with null value as the argument instead.
        """

        def __init__(self):
            super().__init__("removeComment")

        def validate(self):
            """Validates the fields of the request.

            always pass
            """
            pass

        def fileset_change(self):
            return FilesetChange.remove_comment()

File: common/src/main/java/org/apache/gravitino/dto/requests/FilesetUpdateRequest.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.apache.gravitino.dto.requests;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.annotation.JsonSubTypes;
import com.fasterxml.jackson.annotation.JsonTypeInfo;
import com.google.common.base.Preconditions;
import lombok.AllArgsConstructor;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.ToString;
import org.apache.commons.lang3.StringUtils;
import org.apache.gravitino.file.FilesetChange;
import org.apache.gravitino.rest.RESTRequest;

/** Request to update a fileset. */
@JsonIgnoreProperties(ignoreUnknown = true)
@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, include = JsonTypeInfo.As.PROPERTY)
@JsonSubTypes({
  @JsonSubTypes.Type(value = FilesetUpdateRequest.RenameFilesetRequest.class, name = "rename"),
  @JsonSubTypes.Type(
      value = FilesetUpdateRequest.UpdateFilesetCommentRequest.class,
      name = "updateComment"),
  @JsonSubTypes.Type(
      value = FilesetUpdateRequest.RemoveFilesetCommentRequest.class,
      name = "removeComment"),
  @JsonSubTypes.Type(
      value = FilesetUpdateRequest.SetFilesetPropertiesRequest.class,
      name = "setProperty"),
  @JsonSubTypes.Type(
      value = FilesetUpdateRequest.RemoveFilesetPropertiesRequest.class,
      name = "removeProperty")
})
public interface FilesetUpdateRequest extends RESTRequest {

  /**
   * Returns the fileset change.
   *
   * @return the fileset change.
   */
  FilesetChange filesetChange();

  /** The fileset update request for renaming a fileset. */
  @EqualsAndHashCode
  @NoArgsConstructor(force = true)
  @AllArgsConstructor
  @ToString
  class RenameFilesetRequest implements FilesetUpdateRequest {

    @Getter
    @JsonProperty("newName")
    private final String newName;

    /**
     * Returns the fileset change.
     *
     * @return the fileset change.
     */
    @Override
    public FilesetChange filesetChange() {
      return FilesetChange.rename(newName);
    }

    /**
     * Validates the request.
     *
     * @throws IllegalArgumentException if the request is invalid.
     */
    @Override
    public void validate() throws IllegalArgumentException {
      Preconditions.checkArgument(
          StringUtils.isNotBlank(newName), "\"newName\" field is required and cannot be empty");
    }
  }

  /** The fileset update request for updating the comment of a fileset. */
  @EqualsAndHashCode
  @NoArgsConstructor(force = true)
  @AllArgsConstructor
  @ToString
  class UpdateFilesetCommentRequest implements FilesetUpdateRequest {

    @Getter
    @JsonProperty("newComment")
    private final String newComment;

    /**
     * @return The fileset change.
     */
    @Override
    public FilesetChange filesetChange() {
      return FilesetChange.updateComment(newComment);
    }

    /** Validates the fields of the request. Always pass. */
    @Override
    public void validate() throws IllegalArgumentException {}
  }

  /** The fileset update request for setting the properties of a fileset. */
  @EqualsAndHashCode
  @NoArgsConstructor(force = true)
  @AllArgsConstructor
  @ToString
  class SetFilesetPropertiesRequest implements FilesetUpdateRequest {

    @Getter
    @JsonProperty("property")
    private final String property;

    @Getter
    @JsonProperty("value")
    private final String value;

    /**
     * @return The fileset change.
     */
    @Override
    public FilesetChange filesetChange() {
      return FilesetChange.setProperty(property, value);
    }

    /**
     * Validates the request.
     *
     * @throws IllegalArgumentException if the request is invalid.
     */
    @Override
    public void validate() throws IllegalArgumentException {
      Preconditions.checkArgument(
          StringUtils.isNotBlank(property), "\"property\" field is required and cannot be empty");
      Preconditions.checkArgument(value != null, "\"value\" field is required and cannot be null");
    }
  }

  /** The fileset update request for removing the properties of a fileset. */
  @EqualsAndHashCode
  @NoArgsConstructor(force = true)
  @AllArgsConstructor
  @ToString
  class RemoveFilesetPropertiesRequest implements FilesetUpdateRequest {

    @Getter
    @JsonProperty("property")
    private final String property;

    /**
     * @return The fileset change.
     */
    @Override
    public FilesetChange filesetChange() {
      return FilesetChange.removeProperty(property);
    }

    /**
     * Validates the request.
     *
     * @throws IllegalArgumentException if the request is invalid.
     */
    @Override
    public void validate() throws IllegalArgumentException {
      Preconditions.checkArgument(
          StringUtils.isNotBlank(property), "\"property\" field is required and cannot be empty");
    }
  }

  /** The fileset update request for removing the comment of a fileset. */
  @EqualsAndHashCode
  @NoArgsConstructor(force = true)
  @ToString
  class RemoveFilesetCommentRequest implements FilesetUpdateRequest {

    /**
     * @return The fileset change.
     */
    @Override
    public FilesetChange filesetChange() {
      return FilesetChange.updateComment(null);
    }

    /**
     * Validates the request.
     *
     * @throws IllegalArgumentException if the request is invalid.
     */
    @Override
    public void validate() throws IllegalArgumentException {}
  }
}
